Add CouchConfig for coordinating multiple couchdbs

dannyroberts opened the pull request:
This PR
- refactors settings helpers for couch into a class
- adds `CouchConfig` as an interface for accessing couch settings in more intuitive ways,
  as well as figuring out what the equivalent dbs on a remote setup would be
- fixes (and future-proofs) the `copy_domain` bug preventing docs in non-main dbs from being copied

dannyroberts committed:
refactor couch settings helpers into a class

dannyroberts committed:
add CouchConfig for coordinating multiple couchdbs

dannyroberts committed:
have copy_case use CouchConfig

dannyroberts committed:
fix copy_domain for multiple dbs

czue commented:
nice, this looks like it will be very useful

dannyroberts commented:
Error on first build, with the JS tests, is `Server failed to start in allocated time. Exiting.` Ha! @benrudolph in case there's a setting you can tweak.

dannyroberts commented:
@czue yeah I got this little anxious pit in my stomach about reduplicating logic from the settings file for remote commcarehq instances, and this felt much better. (For copy domain, it does make the assumption that the remote commcarehq instance has the same couch settings as the local one—but that seems fine.)

dannyroberts commented:
@esoergel I know some decisions you were making were blocking on whether we could easily patch copy_docs to work on multiple dbs, which is turning out to be relatively straightforward. We *will* need the `domain/docs` view (or its successor `by_domain_doc_type_date/view`) in every db. You can now get all db slugs with `couch_config.all_dbs_by_slug.keys()`, which will make it easier to say "sync this design doc to all dbs" like we'll want to do for these generic ones.

dannyroberts committed:
fix CouchConfig.all_dbs_by_slug

benrudolph commented:
thanks @dannyroberts i think i can bump the time allotted to start the server

esoergel commented on corehq/util/couchdb_management.py:
```diff
@@ -0,0 +1,57 @@
+from django.conf import settings
+
+
+class CouchConfig(object):
```
A docstring would be helpful here.

dannyroberts replied:
Yeah good call

esoergel commented on corehq/util/couchdb_management.py:
```diff
@@ -0,0 +1,57 @@
+    return m
+
+
+couch_config = CouchConfig()
```
Looks like there's some overlap here with https://github.com/dimagi/commcare-hq/blob/master/corehq/util/couch.py#L64-L91
Flagging in case you weren't aware of that stuff and wanted to share code.

dannyroberts replied:
Ah, perfect. `get_document_class_by_name` looks like it does exactly what I want. I'll use that instead, thanks

dannyroberts committed:
fix copy_domain

esoergel commented:
This looks good, @dannyroberts

> You can now get all db slugs with couch_config.all_dbs_by_slug.keys(), which will make it easier to say "sync this design doc to all dbs" like we'll want to do for these generic ones.

Perfect, yeah, I think that that'll be what we want for the "get everything" utilities we have, like copy_domain, staging_replicate, and so forth.

dannyroberts committed:
add docstring to CouchConfig

dannyroberts committed:
tweak get_document_class_by_doc_type

- refactor CouchConfig.get_db_for_doc_type to use it
- rename from get_document_class_by_name
- use doc_type instead of class name
- improve efficiency by memoizing entire mapping

dannyroberts committed:
test and fix CouchConfig

dannyroberts committed:
rearrange util test imports

dannyroberts commented:
@esoergel yeah I ended up fixing copy_domain in this PR, worked like a charm.

Just added fixes and tests for `CouchConfig`, and have tested that `copy_domain` now works locally, across dbs. I'm happy on my end once tests pass.

dannyroberts commented on corehq/apps/domainsync/management/commands/copy_domain.py:
```diff
@@ -91,11 +92,23 @@ class Command(BaseCommand):
+            'receiverwrapper', 'couchlog', 'auditcare', 'fluff-bihar', 'fluff-opm',
+            'fluff-mc', 'fluff-cvsu', 'mvp-indicators', 'm4change',
+            # todo: missing domain/docs, but probably want to add back
+            'meta',
```
This is the only remaining issue. @esoergel I added a note to the cleanup section of [Domain Migration Notes](https://docs.google.com/document/d/1Rc8pYOHkW5QPOCOHqwpooSY3bW45qD8VAOhwCEXdP2o/edit#)

dannyroberts commented:
cc @TylerSheffels buddy

dannyroberts commented:
:white_check_mark:

dannyroberts commented on corehq/util/couchdb_management.py:
```diff
@@ -0,0 +1,63 @@
+    @memoized
+    def all_db_uris_by_slug(self):
+        dbs = self._settings_helper.get_extra_couchdbs()
+        dbs[None] = self.db_uri
```
@TylerSheffels asked what this line does. I guess it's not all that intuitive! `None` is the "slug" for the main db, so this line just adds the main db to the mapping. (`get_extra_couchdbs` only gets the "extra" i.e. non-main dbs.)
